google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/MavenBuildDeploymentSourceType.java

```diff
@@ -67,6 +82,43 @@ protected BeforeRunTask createBuildTask(@NotNull Module module) {
     }
   }
 
+  @NotNull
+  @Override
+  public MavenBuildDeploymentSource load(@NotNull Element tag, @NotNull Project project) {
+    final String moduleName = tag.getAttributeValue(NAME_ATTRIBUTE);
+    Element settings = tag.getChild(DeployToServerRunConfiguration.SETTINGS_ELEMENT);
+
+    if (settings != null) {
+      Module[] modules = ModuleManager.getInstance(project).getModules();
+      Optional<Module> module = Iterables.tryFind(Arrays.asList(modules),
+          new Predicate<Module>() {
+            @Override
+            public boolean apply(Module module) {
+              return module.getName().equals(moduleName);
+            }
+          }
+      );
+
+      String environment = settings.getAttributeValue(
+          AppEngineDeploymentConfiguration.ENVIRONMENT_ATTRIBUTE);
+
+      if (module.isPresent() && environment != null) {
+        return new MavenBuildDeploymentSource(
+            ModulePointerManager.getInstance(project).create(module.get()),
+            project,
+            AppEngineEnvironment.valueOf(environment));
+      }
+    }
+
+    return new MavenBuildDeploymentSource(
```

Comment:
Out of curiosity, when does it happen that we don't know the deployment source environment? Is `settings` expected to be null?